### PR TITLE
Applying what we have learned so far for YieldingContinuation

### PR DESCRIPTION
**Pull request by phausler**
This changes YieldingContinuation into a buffering type since almost every single use case will have some sort of buffering strategy. (buffer sizes of 0 are isomorphic to the dropping behavior previously listed)

Then YieldingContinuation is lifted into a wrapper type to expose it's AsyncSequence-ness and this splits the access points across two types; an Iterator and the YieldingContinuation. The Iterator is not intended to be used in more than one concurrent context (doing so creates a compile failure since the iterator is a mutable type and must mutate to iterate) but the YieldingContinuation type is able to be sendable since it is immutable and only yields values to awaiting consumers (or the consumer's buffer)

### CODE REVIEW

**Comment by phausler** [comment]
@swift-ci please smoke test

**Comment by natecook1000** [comment]
@swift-ci Please smoke test

**Comment by natecook1000** [comment]
@swift-ci Please build toolchain macOS platform

**Comment by phausler** [comment]
@swift-ci please smoke test

**Comment by phausler** [comment]
@swift-ci Please build toolchain macOS platform

**Comment by phausler** [comment]
@swift-ci please smoke test

**Comment by phausler** [comment]
@swift-ci please smoke test

**Comment by phausler** [comment]
@swift-ci please smoke test linux

**Comment by phausler** [comment]
@swift-ci please smoke test windows

**Comment by natecook1000** [comment]
@swift-ci Please build toolchain macOS platform

**Comment by phausler** [comment]
@swift-ci please smoke test

**Comment by phausler** [comment]
@swift-ci please smoke test

**Comment by phausler** [comment]
@swift-ci please smoke test

**Comment by phausler** [comment]
@swift-ci please smoke test linux

**Comment by phausler** [comment]
@swift-ci please smoke test macOS

**Comment by phausler** [comment]
@swift-ci please smoke test linux

**Comment by phausler** [comment]
@swift-ci please smoke test macOS

**Comment by shahmishal** [comment]
@swift-ci please smoke test macOS

**Comment by shahmishal** [comment]
@swift-ci please smoke test linux

**Comment by phausler** [comment]
@swift-ci please smoke test

**Comment by phausler** [comment]
@swift-ci please smoke test

**Comment by phausler** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
> This changes YieldingContinuation into a buffering type since almost every single use case will have some sort of buffering strategy. (buffer sizes of 0 are isomorphic to the dropping behavior previously listed)

Told you so... 😉 

Will review later today.

File: stdlib/public/Concurrency/Series.swift [hunk 1]
```diff
@@ -0,0 +1,638 @@
...
+  /// The maximum number of pending elements limited by dropping the oldest
+  /// value when a new value comes in if the buffer would excede the limit
+  /// placed upon it. By default this limit is unlimited.
+  ///
+  /// The build closure passes in a Continuation which can be used in
+  /// concurrent contexts. It is thread safe to send and finish; all calls are
+  /// to the continuation are serialized, however calling this from multiple
+  /// concurrent contexts could result in out of order delivery.
+  public init(
+    buffering: Element.Type = Element.self,
```
Comment (natecook1000) [inline stdlib/public/Concurrency/Series.swift:409]: I think we were going to remove the default value here — looks like you took it out of `ThrowingSeries` already.

Comment (phausler) [reply stdlib/public/Concurrency/Series.swift:409]: I am really torn on that; it really ends up being quite nice to write the return values... e.g.

```swift
var someSeriesOfThings: Series<Thing> {
   Series { continuation in
      ...
   }
}
```

**Comment by phausler** [comment]
@swift-ci please smoke test

**Comment by phausler** [comment]
@swift-ci please smoke test

**Comment by phausler** [comment]
@swift-ci please smoke test macOS

**Comment by phausler** [comment]
@swift-ci please smoke test

File: stdlib/public/Concurrency/AsyncStreamBuffer.swift [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+    let handler = state.onTermination
+    state.onTermination = nil
+    unlock()
+    
+    handler?.cancel() // handler must be invoked before yielding nil for termination
+
+    yield(nil)
+  }
+
+  private func enqueue(_ value: __owned Element?) -> TerminationHandler? {
```
Comment (natecook1000) [inline stdlib/public/Concurrency/AsyncStreamBuffer.swift:146]: Could you add a comment on `enqueue` and `terminate` that they can only be called inside a lock?

Comment (phausler) [reply stdlib/public/Concurrency/AsyncStreamBuffer.swift:146]: will do.

Comment (phausler) [reply stdlib/public/Concurrency/AsyncStreamBuffer.swift:147]: fixed, also while auditing that I caught an issue where the termination handler was called after terminal events were propagated.

**Comment by phausler** [comment]
@swift-ci please smoke test

**Comment by phausler** [comment]
@swift-ci please smoke test linux

File: stdlib/public/Concurrency/Deque.swift [hunk 3]
```diff
@@ -0,0 +1,502 @@
...
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Swift
+
+@available(SwiftStdlib 5.5, *)
+@frozen
+@usableFromInline
+struct _Deque<Element> {
```
Comment (DougGregor) [inline stdlib/public/Concurrency/Deque.swift:18]: This is a lot of ABI surface area to take on. Is this the Deque implementation for the Swift Collections package, or something bespoke?

File: stdlib/public/Darwin/Dispatch/Schedulers+DispatchQueue.swift [hunk 4]
```diff
@@ -1,3 +1,4 @@
+#if false
```
Comment (DougGregor) [inline stdlib/public/Darwin/Dispatch/Schedulers+DispatchQueue.swift:1]: This shouldn't be here.

**Review by DougGregor — APPROVED** [review]
I'm okay with this, but have some questions, and see the two places I marked where some build cruft worked its way into the PR.

File: stdlib/public/Concurrency/AsyncThrowingStream.swift [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    let storage: _Storage
+
+    /// Resume the task awaiting the next iteration point by having it return
+    /// nomally from its suspension point or buffer the value if no awaiting
+    /// next iteration is active.
+    ///
+    /// - Parameter value: The value to yield from the continuation.
+    ///
+    /// This can be called more than once and returns to the caller immediately
+    /// without blocking for any awaiting consuption from the iteration.
```
Comment (heckj) [inline stdlib/public/Concurrency/AsyncThrowingStream.swift:32]:
```suggestion
    /// without blocking for any awaiting consumption from the iteration.
```

File: stdlib/public/Concurrency/AsyncThrowingStream.swift [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+  /// Construct an iterator.
+  public func makeAsyncIterator() -> Iterator {
+    return Iterator(produce: produce)
+  }
+}
+
+@available(SwiftStdlib 5.5, *)
+extension AsyncThrowingStream.Continuation {
+  /// Resume the task awaiting the next iteration point by having it return
+  /// nomally from its suspension point or buffer the value if no awaiting
```
Comment (heckj) [inline stdlib/public/Concurrency/AsyncThrowingStream.swift:122]:
```suggestion
  /// normally from its suspension point or buffer the value if no awaiting
```

File: stdlib/public/Concurrency/AsyncThrowingStream.swift [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    switch result {
+      case .success(let val):
+        storage.yield(val)
+      case .failure(let err):
+        storage.finish(throwing: err)
+    }
+  }
+
+  /// Resume the task awaiting the next iteration point by having it return
+  /// nomally from its suspension point or buffer the value if no awaiting
```
Comment (heckj) [inline stdlib/public/Concurrency/AsyncThrowingStream.swift:141]:
```suggestion
  /// normally from its suspension point or buffer the value if no awaiting
```

File: stdlib/public/Concurrency/Deque.swift (hunk 3 above)
Comment (phausler) [reply stdlib/public/Concurrency/Deque.swift:18]: this is 100% taken from the Swift Collections package, and nothing should be exposed for this to external callers

**Comment by phausler** [comment]
@swift-ci please smoke test

**Comment by phausler** [comment]
@swift-ci please smoke test macOS

File: stdlib/public/Concurrency/AsyncStream.cpp [hunk 8]
```diff
@@ -0,0 +1,39 @@
+//===--- AsyncStream.cpp - Multi-resume locking interface -----------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
```
Comment (donnywals) [inline stdlib/public/Concurrency/AsyncStream.cpp:5]: // Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors

**Pull request merged**